### DISCUSSION TITLE
Can I have manual updates with Conveyor?

### DISCUSSION THREAD

**Discussion by MohamedRejeb** [hydraulic-software/conveyor#106] [Q&A] [answered]
Hi,
I'm looking for something similar to what we have in IntellijIDEA, the user can check for updates, and he can manually download the update when he wants. Can we do something like that with conveyor and Compose?

Thanks!

**Comment by mikehearn** [comment]
Hi,

Conveyor has all the foundations for that, but at the moment the API needed to trigger manual updates isn't publicly exposed. If you're going to become a commercial user, we can send you the docs needed to use it anyway. We want to make it public soon but have been working on higher priority things. It does work, this is how Conveyor itself is updated, but it needs a bit of polish.

Could you help us understand why you need manually triggered updates? Even with our API, it doesn't offer any UI widgets for this so you'd have to design your own notifications and workflow around it.

**Reply by acarlsen** [reply]
I would also very much like this.  Automatic updates is fine while the users licensing covers updates, but lets say a specific license only has updates for 2 years, then I don't won't to auto-update the user to a newer version where the license may be invalid (and rollback may be hard or impossible). 

Then I want custom logic to trigger the updated based on if the new version is covered by license, else I want to inform the user of consequences of updating (needing to purchase extensions to license).

**Reply by MohamedRejeb** [reply]
I'm having a similar case to @acarlsen, I don't want users with expired license to get the updates. I don't want to remove background updates, but also I want to provide an option to check for updates and manual updates. Because users are used to it on most desktop apps.
Also, is the standard plan considered commercial user?

**Reply by mikehearn** [reply]
Thanks everyone. This is helpful. It's good I asked: the issue here is that you want to turn off background updates after some external event happens like the user not paying. The current API doesn't really do this. It only lets you check for an update manually. The closest equivalent would be to turn off automatic polling and then trigger the update checks yourself, but of course it means that the app will try to update when the user has just started it and is probably trying to do something else. That's why we have background updates (on Windows at least, maybe in future on macOS or in the app stores).

We will prioritize publishing the API. But it won't quite fit this use case, so, we need to do some more design work here. Ideally the app would update quietly (like Chrome) and users with expired licenses would simply stay on the same version forever, or until they renew. The app can then expose this fact via the UI if it wishes to, but the user would not be expected to "drive" the update process. Especially ideal if the user cannot download the new version at all.

One thing we've considered is a special server (or service) that lets you control the version that each and every user is updated to. This can then be used to integrate with a licensing system, or to do canary rollouts, or to roll individual users back if they file support tickets indicating a regression. Probably there'd be some API that lets you learn what a user's install UUID is, and a server side API to control what version each UUID gets. Would this be of interest?

**Reply by acarlsen** [reply]
I would like to control which users get updates based on license, but I don't want another server/service to be dependent on, other than the licensing server.
 
I just wish for programmatically to disable all auto updating if there is a need to and then I can inform the user about any updates however I deem necessary in the app. Similar to how the Forklift app on Mac does it: allows updates while you have license that allow it, but shows you a warning if you don't (and doesn't allow the update).

Of course there may be some framework limitations that make this hard/impossible, I don't really now much about how the updates process works on the different platforms.

If not possible, then maybe have an option to build an app with auto-updating completely disabled (but the option to trigger an update programmatically when app is open if needed). Don't know if that makes it easier..

**Reply by mikehearn** [reply]
We can have auto updating be disabled and only let the app trigger it. That's the model Conveyor itself uses.

One minor issue is that on Windows the installer always installs the latest version of the app (the update and install codepaths are the same). So if a user stops paying, stops updating, and then later reinstalls the app on e.g. a new machine, they will get a newer version that they aren't entitled to run. Other than having the app force a downgrade when it first runs, I don't see a good solution to that without update server/license server integration.

**Reply by acarlsen** [reply]
That model would work for me.

I guess on windows you could direct users with expired licenses to use the zip/msix file if they want to re-install the old version.

**Reply by MohamedRejeb** [reply]
Any updates on this ?

**Reply by mikehearn** [reply]
We are making progress on this feature but can't commit to any timeline, sorry.

**Comment by mikehearn** [comment, marked as answer]
We've just launched Conveyor 15, which adds a new control API! With this you can implement your own update UI. Let us know what you think:

https://conveyor.hydraulic.dev/15.0/release-notes/

**Reply by MohamedRejeb** [reply]
Great! I will give it a try.
